# MACHINE LEARNING

### LINEAR REGRESSION

Machine Learning merupakan salah satu cabang ilmu Kecerdasan Buatan (Artificial Intellegence). Machine learning membahas mengenai pembuatan suatu sistem yang dimana komputer belajar dari data yang tersedia.

Aplikasi Machine Learning :

1. Recognition (Pengenalan) : face, object, speech, gait, handwritten <br>
2. Detection (Deteksi), identification : object, fraud (credit card), spam email, disease <br>
3. Recommender problems (Sistem Rekomendasi) : the most liked movies, products, and etc <br>
4. Prediction (Prediksi) : stock prices, housing <br>
5. dll

Learning :
1. Supervised <br>
   Terdapat data dan target / label
2. Unsupervised <br>
   Berupa pengelompokan data / filtering data berdasarkan fitur yang sama

## Supervised Learning - Regression

Bertujuan untuk prediksi / menghasilkan real value (continuos) yang outputnya belum tentu atau tak terbatas <br><br>

### Linear Regression

Linear Regression mengasumsikan relasi antara input dan output. Dibagi menjadi 2 jenis, yaitu :
![](https://4.bp.blogspot.com/-gNcMdvyWDw0/W5Ukses0RuI/AAAAAAAAFfs/B3LJBD5AkQ0NpwC54IKC06yvdzs58C1CwCLcBGAs/s1600/Capture.PNG)
<br>
f(x;w) atau y merupakan output yang dihasilkan, x adalah input atau fitur/atribut, dan w merupakan parameter yang akan dicari. w0 disebut juga dengan bias atau intercept. Untuk simple linear regression hanya ada 1 dimensi, sedangkan multiple linear regression memiliki lebih dari 1 dimensi (d dimensi).

#### Simple Linear Regression
Contohnya adalah prediksi harga rumah berdasarkan luas area. Datanya adalah sebagai berikut :<br>
<img style="float:left" src="https://4.bp.blogspot.com/-vu8dMOCYFFI/W5UnISjoJWI/AAAAAAAAFf4/FTUY8rFpmDEiVZfRDQj6TwxuZu4vGKxZgCLcBGAs/s1600/Capture.PNG">

<em> Proses Simple Linear Regression </em>

- Learning (Pembelajaran)

In [7]:
from sklearn.linear_model import LinearRegression
X=[[2104],[1416],[1539],[852],[1940]]
Y=[[400],[232],[315],[178],[240]]
model = LinearRegression()
model.fit(X, Y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

- Parameter

In [8]:
print(model.coef_)
print(model.intercept_)

[[ 0.13388318]]
[ 62.77663135]


- Testing (Uji Coba)

In [9]:
model.predict(1250)

array([[ 230.13060588]])

Dari hasil percobaan di atas, didapatkan hasil pencarian parameternya adalah w0 = 0.13388318 dan w1 = 62.77663135. Ketika dilakukan prediksi dengan input luas area (x) = 1250 didapatkan hasil (y) = 230.13060588 <br>
Percobaan di atas menggunakan library yang disediakan oleh python. Sedangkan untuk melakukan percobaan menggunakan perhitungan manual, dapat menggunakan rumus berikut :<br>
![](https://4.bp.blogspot.com/-Rj-bd7nuRdI/W5UvAH_t4EI/AAAAAAAAFgc/VsWX21sMRKwBvgCFNpFkePHsywa-AfCdwCLcBGAs/s1600/Capture.PNG) <br>
Keterangan :
1. Rumus [1] digunakan untuk mencari nilai variance / persimpangan
2. Rumus [2] digunakan untuk mencari nilai covariance / korelasi
3. Rumus [3] digunakan untuk mencari nilai koefisien
4. Rumus [4] digunakan untuk mencari nilai intercept
<br>

#### Multivariate Linear Regression


Terdapat cara lain untuk mendapatkan nilai parameter koefisien maupun intercept, yaitu menggunakan <em>Gradient Descent</em>
Rumus yang digunakan adalah sebagai berikut : 
![](https://1.bp.blogspot.com/-72aufmf7PbU/W5Uyt3tDJxI/AAAAAAAAFgo/4bacXyPf6IQ-PfXF4rmmlSsy5PSYBB4xwCLcBGAs/s1600/Capture.PNG)
<br> Percobaan menggunakan library python

In [11]:
from numpy.linalg import inv
from numpy import dot, transpose
X=[[1,2104],[1,1416],[1,1539],[1,852],[1,1940]]
Y=[[400],[232],[315],[178],[240]]
print (dot(inv(dot(transpose(X), X)), dot(transpose(X), Y)))

[[ 62.77663135]
 [  0.13388318]]


Didapatkan nilai parameter koefisien = 62.77663135 dan intercept = 0.13388318